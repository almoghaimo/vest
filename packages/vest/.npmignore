# Autogenerated section. Do not edit manually.
node_modules
src
!types/
!dist/
tsconfig.json
!promisify/
!parser/
!enforce@schema/
!enforce@isEmail/
!enforce@compounds/
!enforce@compose/
!classnames/

# Manual Section. Edit at will.