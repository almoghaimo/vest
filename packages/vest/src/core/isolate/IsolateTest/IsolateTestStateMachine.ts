import { StateMachine, TStateMachine } from 'vest-utils';

export enum TestStatus {
  UNTESTED = 'UNTESTED',
  SKIPPED = 'SKIPPED',
  FAILED = 'FAILED',
  WARNING = 'WARNING',
  PASSING = 'PASSING',
  PENDING = 'PENDING',
  CANCELED = 'CANCELED',
  OMITTED = 'OMITTED',
}

export enum TestAction {
  RESET = 'RESET',
}

type TAction = TestAction | TestStatus;

export function createTestStateMachine() {
  return StateMachine<TestStatus, TAction>(machine);
}

/* eslint-disable sort-keys */
const machine: TStateMachine<TestStatus, TAction> = {
  initial: TestStatus.UNTESTED,
  states: {
    '*': {
      [TestStatus.OMITTED]: TestStatus.OMITTED,
      [TestAction.RESET]: TestStatus.UNTESTED,
    },
    [TestStatus.UNTESTED]: {
      [TestStatus.CANCELED]: TestStatus.CANCELED,
      [TestStatus.FAILED]: TestStatus.FAILED,

      [TestStatus.PASSING]: TestStatus.PASSING,
      [TestStatus.PENDING]: TestStatus.PENDING,
      [TestStatus.SKIPPED]: TestStatus.SKIPPED,
      [TestStatus.WARNING]: TestStatus.WARNING,
    },
    [TestStatus.PENDING]: {
      [TestStatus.CANCELED]: TestStatus.CANCELED,
      [TestStatus.FAILED]: TestStatus.FAILED,

      [TestStatus.PASSING]: TestStatus.PASSING,
      [TestStatus.SKIPPED]: [
        TestStatus.SKIPPED,
        (force?: boolean) => force === true,
      ],
      [TestStatus.WARNING]: TestStatus.WARNING,
    },
    [TestStatus.SKIPPED]: {},
    [TestStatus.FAILED]: {},
    [TestStatus.WARNING]: {},
    [TestStatus.PASSING]: {},
    [TestStatus.CANCELED]: {},
    [TestStatus.OMITTED]: {},
  },
};
/* eslint-enable sort-keys */
