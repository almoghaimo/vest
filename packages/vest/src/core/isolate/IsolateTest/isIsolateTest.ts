import { invariant } from 'vest-utils';
import { Isolate } from 'vestjs-runtime';

import type { IsolateTest } from 'IsolateTest';
import { TFieldName, TGroupName } from 'SuiteResultTypes';
import { VestIsolateType } from 'VestIsolateType';

export function isIsolateTest(isolate?: Isolate): isolate is IsolateTest {
  return isolate?.type === VestIsolateType.Test;
}

export function isIsolateTestX(
  isolate: Isolate
): asserts isolate is IsolateTest {
  invariant(isIsolateTest(isolate));
}

export function castIsolateTest<
  F extends TFieldName = TFieldName,
  G extends TGroupName = TGroupName
>(isolate: Isolate): IsolateTest<F, G> {
  isIsolateTestX(isolate);

  return isolate as IsolateTest<F, G>;
}
