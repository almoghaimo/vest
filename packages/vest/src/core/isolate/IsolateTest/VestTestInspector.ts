import { isPromise } from 'vest-utils';

import type { IsolateTest } from 'IsolateTest';
import { TestStatus } from 'IsolateTestStateMachine';
import { TestSeverity } from 'Severity';

export class VestTestInspector {
  static warns(test: IsolateTest): boolean {
    return test.severity === TestSeverity.Warning;
  }

  static isPending(test: IsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.PENDING);
  }

  static isOmitted(test: IsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.OMITTED);
  }

  static isUntested(test: IsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.UNTESTED);
  }

  static isFailing(test: IsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.FAILED);
  }

  static isCanceled(test: IsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.CANCELED);
  }

  static isSkipped(test: IsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.SKIPPED);
  }

  static isPassing(test: IsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.PASSING);
  }

  static isWarning(test: IsolateTest): boolean {
    return VestTestInspector.statusEquals(test, TestStatus.WARNING);
  }

  static hasFailures(test: IsolateTest): boolean {
    return (
      VestTestInspector.isFailing(test) || VestTestInspector.isWarning(test)
    );
  }

  static isNonActionable(test: IsolateTest): boolean {
    return (
      VestTestInspector.isSkipped(test) ||
      VestTestInspector.isOmitted(test) ||
      VestTestInspector.isCanceled(test)
    );
  }

  static isTested(test: IsolateTest): boolean {
    return (
      VestTestInspector.hasFailures(test) || VestTestInspector.isPassing(test)
    );
  }

  static awaitsResolution(test: IsolateTest): boolean {
    // Is the test in a state where it can still be run, or complete running
    // and its final status is indeterminate?
    return (
      VestTestInspector.isSkipped(test) ||
      VestTestInspector.isUntested(test) ||
      VestTestInspector.isPending(test)
    );
  }

  static isAsyncTest(test: IsolateTest): boolean {
    return isPromise(test.asyncTest);
  }

  static statusEquals(test: IsolateTest, status: TestStatus): boolean {
    return test.stateMachine.getState() === status;
  }
}
