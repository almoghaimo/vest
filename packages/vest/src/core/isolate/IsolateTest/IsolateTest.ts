import { Maybe, seq } from 'vest-utils';
import { Isolate, IsolateKey, IsolateMutator } from 'vestjs-runtime';

import { IsolateTestReconciler } from 'IsolateTestReconciler';
import { createTestStateMachine } from 'IsolateTestStateMachine';
import { TestSeverity } from 'Severity';
import { TFieldName, TGroupName } from 'SuiteResultTypes';
import { TestFn, AsyncTest } from 'TestTypes';
import { VestIsolateType } from 'VestIsolateType';
import { VestTestInspector } from 'VestTestInspector';

type IsolateTestInput = {
  message?: string;
  groupName?: string;
  fieldName: TFieldName;
  testFn: TestFn;
  key?: IsolateKey;
};

export class IsolateTest<
  F extends TFieldName = TFieldName,
  G extends TGroupName = TGroupName
> extends Isolate {
  children = null;
  fieldName: F;
  testFn: TestFn;
  groupName?: G;
  message?: Maybe<string>;
  asyncTest?: AsyncTest;
  id = seq();
  severity = TestSeverity.Error;
  type = VestIsolateType.Test;
  stateMachine = createTestStateMachine();

  static reconciler = IsolateTestReconciler;

  constructor({
    fieldName,
    testFn,
    message,
    groupName,
    key = null,
  }: IsolateTestInput) {
    super();

    this.fieldName = fieldName as F;
    this.testFn = testFn;

    if (groupName) {
      this.groupName = groupName as G;
    }

    if (message) {
      this.message = message;
    }

    IsolateMutator.setKey(this, key);
  }

  valueOf(): boolean {
    return !VestTestInspector.isFailing(this);
  }
}
