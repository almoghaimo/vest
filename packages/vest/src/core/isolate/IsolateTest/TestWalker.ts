import { Nullable } from 'vest-utils';
import { Walker, VestRuntime, Isolate } from 'vestjs-runtime';

import { IsolateTest } from 'IsolateTest';
import { TFieldName, TGroupName } from 'SuiteResultTypes';
import { VestTestInspector } from 'VestTestInspector';
import { VestTestMutator } from 'VestTestMutator';
import { castIsolateTest, isIsolateTest, isIsolateTestX } from 'isIsolateTest';
import matchingFieldName from 'matchingFieldName';

type MaybeRoot = Nullable<Isolate>;

export class TestWalker {
  static defaultRoot = VestRuntime.useAvailableRoot;

  static hasNoTests(root: MaybeRoot = TestWalker.defaultRoot()): boolean {
    if (!root) return true;
    return !Walker.has(root, isIsolateTest);
  }

  static someIncompleteTests(
    predicate: (test: IsolateTest) => boolean,
    root: MaybeRoot = TestWalker.defaultRoot()
  ): boolean {
    if (!root) return false;
    return Walker.some(
      root,
      isolate => {
        isIsolateTestX(isolate);

        return VestTestInspector.isPending(isolate) && predicate(isolate);
      },
      isIsolateTest
    );
  }

  static someTests(
    predicate: (test: IsolateTest) => boolean,
    root: MaybeRoot = TestWalker.defaultRoot()
  ): boolean {
    if (!root) return false;
    return Walker.some(
      root,
      isolate => {
        isIsolateTestX(isolate);

        return predicate(isolate);
      },
      isIsolateTest
    );
  }

  static everyTest(
    predicate: (test: IsolateTest) => boolean,
    root: MaybeRoot = TestWalker.defaultRoot()
  ): boolean {
    if (!root) return false;
    return Walker.every(
      root,
      isolate => {
        isIsolateTestX(isolate);

        return predicate(isolate);
      },
      isIsolateTest
    );
  }

  static walkTests<F extends TFieldName, G extends TGroupName>(
    callback: (test: IsolateTest<F, G>, breakout: () => void) => void,
    root: MaybeRoot = TestWalker.defaultRoot()
  ): void {
    if (!root) return;
    Walker.walk(
      root,
      (isolate, breakout) => {
        callback(castIsolateTest(isolate), breakout);
      },
      isIsolateTest
    );
  }

  static hasRemainingTests(fieldName?: TFieldName): boolean {
    return TestWalker.someIncompleteTests(testObject => {
      if (fieldName) {
        return matchingFieldName(testObject, fieldName);
      }
      return true;
    });
  }

  static pluckTests(
    predicate: (test: IsolateTest) => boolean,
    root: MaybeRoot = TestWalker.defaultRoot()
  ): void {
    if (!root) return;
    Walker.pluck(
      root,
      isolate => {
        isIsolateTestX(isolate);

        return predicate(isolate);
      },
      isIsolateTest
    );
  }

  static resetField(fieldName: TFieldName): void {
    TestWalker.walkTests(testObject => {
      if (matchingFieldName(testObject, fieldName)) {
        VestTestMutator.reset(testObject);
      }
    }, TestWalker.defaultRoot());
  }

  static removeTestByFieldName(
    fieldName: TFieldName,
    root: MaybeRoot = TestWalker.defaultRoot()
  ): void {
    TestWalker.pluckTests(testObject => {
      return matchingFieldName(testObject, fieldName);
    }, root);
  }
}
