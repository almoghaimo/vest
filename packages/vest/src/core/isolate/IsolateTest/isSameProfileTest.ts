import { IsolateTest } from 'IsolateTest';
import matchingFieldName from 'matchingFieldName';

export function isSameProfileTest(
  testObject1: IsolateTest,
  testObject2: IsolateTest
): boolean {
  return (
    matchingFieldName(testObject1, testObject2.fieldName) &&
    testObject1.groupName === testObject2.groupName &&
    testObject1.key === testObject2.key
  );
}
