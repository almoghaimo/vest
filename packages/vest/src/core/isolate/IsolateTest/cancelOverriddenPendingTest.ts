import { IsolateTest } from 'IsolateTest';
import { isSameProfileTest } from 'isSameProfileTest';

export default function cancelOverriddenPendingTest(
  prevRunTestObject: IsolateTest,
  currentRunTestObject: IsolateTest
): void {
  if (
    currentRunTestObject !== prevRunTestObject &&
    isSameProfileTest(prevRunTestObject, currentRunTestObject) &&
    prevRunTestObject.isPending()
  ) {
    prevRunTestObject.cancel();
  }
}
