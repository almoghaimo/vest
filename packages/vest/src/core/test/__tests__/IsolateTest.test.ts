import wait from 'wait';

import { TestPromise } from '../../../../testUtils/testPromise';

import { IsolateTest } from 'IsolateTest';
import { VestTestInspector } from 'VestTestInspector';
import * as vest from 'vest';

const fieldName = 'unicycle';
const message = 'I am Root.';

describe('IsolateTest', () => {
  let testObject: IsolateTest;

  beforeEach(() => {
    testObject = new IsolateTest({
      fieldName,
      testFn: jest.fn(),
      message,
    });
  });

  test('TestObject constructor', () => {
    expect(testObject).toMatchSnapshot();
  });

  it('Should have a unique id', () => {
    Array.from(
      { length: 100 },
      () =>
        new IsolateTest({
          fieldName,
          testFn: jest.fn(),
          message,
        })
    ).reduce((existing, { id }) => {
      expect(existing.has(id)).toBe(false);
      existing.add(id);
      return existing;
    }, new Set<string>());
  });

  describe('testObject.warn', () => {
    it('Should mark the test as warning', () => {
      expect(VestTestInspector.warns(testObject)).toBe(false);
      testObject.warn();
      expect(VestTestInspector.warns(testObject)).toBe(true);
      expect(testObject).toMatchSnapshot();
    });
  });

  describe('testObject.fail', () => {
    beforeEach(() => {
      jest.resetModules();

      const { IsolateTest } = require('IsolateTest'); // eslint-disable-line @typescript-eslint/no-var-requires
      testObject = new IsolateTest(fieldName, jest.fn(), { message });
    });

    afterEach(() => {
      jest.resetAllMocks();
    });

    it('Should set status to failed', () => {
      expect(VestTestInspector.isFailing(testObject)).toBe(false);
      testObject.fail();
      expect(VestTestInspector.isFailing(testObject)).toBe(true);
    });
  });

  describe('testObject.valueOf', () => {
    test('When test did not fail', () => {
      expect(testObject.valueOf()).toBe(true);
    });

    test('When test failed', () => {
      testObject.fail();
      expect(testObject.valueOf()).toBe(false);
    });
  });

  describe('testObject.cancel', () => {
    it('Should set the testObject to cancel', () => {
      let testObject: IsolateTest;
      return TestPromise(done => {
        const suite = vest.create(() => {
          testObject = vest.test('f1', async () => {
            await wait(100);
          });
          vest.test('f2', async () => {
            await wait(100);
          });
          testObject.cancel();
        });
        suite();

        expect(VestTestInspector.isCanceled(testObject)).toBe(true);
        done();
      });
    });

    describe('final statuses', () => {
      let control = jest.fn();
      beforeEach(() => {
        control = jest.fn();
      });
      it('keep status unchanged when `failed`', () => {
        vest.create(() => {
          // async so it is not a final status
          const testObject = vest.test('f1', async () => {
            await wait(100);
          });
          testObject.fail();
          expect(VestTestInspector.isFailing(testObject)).toBe(true);
          testObject.skip();
          expect(VestTestInspector.isSkipped(testObject)).toBe(false);
          expect(VestTestInspector.isFailing(testObject)).toBe(true);
          testObject.cancel();
          expect(VestTestInspector.isCanceled(testObject)).toBe(false);
          expect(VestTestInspector.isFailing(testObject)).toBe(true);
          testObject.setPending();
          expect(VestTestInspector.isPending(testObject)).toBe(false);
          expect(VestTestInspector.isFailing(testObject)).toBe(true);
          control();
        })();
        expect(control).toHaveBeenCalledTimes(1);
      });

      it('keep status unchanged when `canceled`', () => {
        vest.create(() => {
          // async so it is not a final status
          const testObject = vest.test('f1', async () => {
            await wait(100);
          });
          testObject.cancel();
          expect(VestTestInspector.isCanceled(testObject)).toBe(true);
          testObject.fail();
          expect(VestTestInspector.isCanceled(testObject)).toBe(true);
          expect(VestTestInspector.isFailing(testObject)).toBe(false);
          testObject.skip();
          expect(VestTestInspector.isSkipped(testObject)).toBe(false);
          expect(VestTestInspector.isCanceled(testObject)).toBe(true);
          testObject.setPending();
          expect(VestTestInspector.isPending(testObject)).toBe(false);
          expect(VestTestInspector.isCanceled(testObject)).toBe(true);
          control();
        })();
        expect(control).toHaveBeenCalledTimes(1);
      });
    });
  });
});
