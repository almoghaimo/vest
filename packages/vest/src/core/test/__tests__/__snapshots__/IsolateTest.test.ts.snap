// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IsolateTest TestObject constructor 1`] = `
IsolateTest {
  "children": null,
  "fieldName": "unicycle",
  "id": "0",
  "key": null,
  "keys": {},
  "message": "I am Root.",
  "parent": null,
  "severity": "error",
  "stateMachine": {
    "getState": [Function],
    "transition": [Function],
  },
  "testFn": [MockFunction],
  "type": "Test",
}
`;

exports[`IsolateTest testObject.warn Should mark the test as warning 1`] = `
IsolateTest {
  "children": null,
  "fieldName": "unicycle",
  "id": "102",
  "key": null,
  "keys": {},
  "message": "I am Root.",
  "parent": null,
  "severity": "warning",
  "stateMachine": {
    "getState": [Function],
    "transition": [Function],
  },
  "testFn": [MockFunction],
  "type": "Test",
}
`;
